parser-typechecker/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -415,23 +427,24 @@ loop = do
         stepManyAt = Unison.Codebase.Editor.HandleInput.stepManyAt inputDescription
         stepManyAtM = Unison.Codebase.Editor.HandleInput.stepManyAtM inputDescription
         updateAtM = Unison.Codebase.Editor.HandleInput.updateAtM inputDescription
-        manageLinks :: Var v0
-                    => [(Path', NameSegment.HQSegment)]
-                    -> (Path', NameSegment.HQSegment)
+        manageLinks :: [(Path', NameSegment.HQSegment)]
+                    -> HQ.HashQualified
                     -> (forall r. Ord r
                         => (r, Reference, Reference)
                         ->  Branch.Star r NameSegment
                         ->  Branch.Star r NameSegment)
-                    -> MaybeT (StateT (LoopState m v0) (F m (Either Event Input) v0)) ()
-        manageLinks srcs mdValue op = do
-          let srcle = toList . getHQ'Terms =<< srcs
-              srclt = toList . getHQ'Types =<< srcs
-              mdValuel = toList (getHQ'Terms mdValue)
+                    -> MaybeT (StateT (LoopState m v) (F m (Either Event Input) v)) ()
+        manageLinks srcs mdValue2 op = do
+          let !srcle = toList . getHQ'Terms =<< srcs
+              !srclt = toList . getHQ'Types =<< srcs
+          mdValuel <- toList <$> getHQTerms mdValue2
+          names0 <- basicPrettyPrintNames0
+          ppe <- PPE.suffixifiedPPE <$> prettyPrintEnvDecl (Names names0 mempty)
           case (srcle, srclt, mdValuel) of
-            (srcle, srclt, [Referent.Ref mdValue]) -> do
+            (srcle, srclt, [r@(Referent.Ref mdValue)]) -> do
               mdType <- eval $ LoadTypeOfTerm mdValue
               case mdType of
-                Nothing -> respond $ LinkFailure input
+                Nothing -> respond $ MetadataMissingType ppe r
```

Comment:
We previously just had `LinkFailure`, which I split into `MetadataMissingType` and `MetadataAmbiguous` messages.